[MVC-Step 2] 미스터코 미션 제출합니다.

eunsukko opened the pull request:
써머 안녕하세요..! 

일단 step1 때 주셨던 피드백 반영한다고 해보았습니다. (테스트를 전반적으로 살펴보라고 하셨는데.. AnnotationRequestMapping 위주가 된것도 같네요...ㅎ 다른 부분을 말씀해주셨던 걸까요??)
또한 말씀해주신 RequestMethod.ALL 도 없는 방향이 맞다고 생각이 되어서 다른 RequestMethod 를 모두 추가하는 방식으로 변경하였습니다.

(JDBC 리뷰중이실것같은데 뜬금없는 MVC 요청일까도 싶네요... ㅠㅎ)

감사합니다..!!

eunsukko committed:
[refactor] ManualHandlerMapping 클래스 이동 (slipp -> nextstep/mvc)

eunsukko committed:
[refactor] testdata package 생성

- 왜 변경(혹은 추가)이 있었나요?
	- mvc 테스트에서 테스트를 위한 클래스들이 필요했음
	- 그런데 테스트코드와 함께 있으니 복잡함
		- 테스트코드 vs 테스트를 위한 코드
- 무엇이 변경되었나요?
	- testdata package 생성
		- 테스트를 위한 코드를 모아놓는 역할

eunsukko committed:
[refactor] DispatcherServlet 리팩토링

- 왜 변경(혹은 추가)이 있었나요?
	- 기존에는 HandlerAdapter 가 Handler 역할을 정확하게 대체하지 못했음
		- 매번 Adapter 와 Hanlder 역할을 하는 객체를 조합해서 사용해야 했음
	- 사용하는 곳에서 Handler 만 신경쓸수 있도록 하고 싶었음
- 무엇이 변경되었나요?
	- HandlerFactory 를 통한 생성
		- 반환된 Handler 는 Adapter 가 적용되어 있음

eunsukko committed:
[refactor] AnnotationHandlerMapping 리팩토링 (역할을 위임)

- 왜 변경(혹은 추가)이 있었나요?
	- AnnotationHandlerMapping 에서 하는 역할이 많음
		- @Controller 달린 클래스 찾기
		- 각 클래스에서 @RequestMapping 달린 메소드 찾기
		- 각 메소드에서 정보를 추출해서 mapping 에 추가
			- 올바른 RequestMapping 정보인지 검증
			- 올바른 HandlerExecution 시그니처인지 검증
		- 예외 처리
	- 테스트 하기에도 복잡함
		- 각각의 역할이 조합이 되기에 경우의 수가 많아짐

- 무엇이 변경되었나요?
	- AnnotationHandlerMapping 에서는 흐름과 예외를 제어하도록
		- 메소드에서 추출한 정보를 가지고 HandlerKey, HandlerExecution 을 생성 및 검증 하는 부분을 위임
	- RequestMethod.ALL 을 제거
		- RequestMethod.ALL 을 사용했던 곳에서 모든 종류의 RequestMethod 들을 등록

eunsukko committed:
[feat] UserApiController 추가(create, show, update)

- JsonView 구현
- UserApiController 구현
	- 유저 추가
	- 유저 반환
	- 유저 정보 수정

jihan805 commented on nextstep-mvc/src/main/java/nextstep/utils/JsonUtils.java:
```diff
@@ -19,4 +26,18 @@
+//            objectMapper.writeValue(outputStream, instance);
+//
+//            return outputStream.toString();
+        } catch (IOException e) {
```
주석 처리한 objectMapper.writeValue()를 이용해 더 깔끔하게 구현할 수 있을것 같아요!

eunsukko replied:
말씀하신 깔끔한 구현이 무엇일까 생각해보다가 이런식으로 적용해보았습니다!

``` java
    // toJson
    public static String toJson(Object instance) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();

        writeJsonTo(out, instance);
        return out.toString();
    }

    private static void writeJsonTo(OutputStream out, Object instanceTobeJson) {
        try {
            objectMapper.writeValue(out, instanceTobeJson);
        } catch (IOException e) {
            throw new ObjectMapperException(e);
        }
    }
```
string 말고도 json 을 써줄 수 있는 애들이 많을 수도 있겠다 싶어서 이렇게 만들었습니다.
(나중에 파일 같은 곳에 써도.. 비슷하게 확장 할 수 있도록이요..!)

다른 의도이셨으면 알려주셔요~ㅎ

jihan805 commented on nextstep-mvc/src/test/java/nextstep/db/DataBase.java:
```diff
@@ -1,7 +1,7 @@
 
 import com.google.common.collect.Maps;
-import nextstep.mvc.tobe.User;
+import nextstep.mvc.tobe.testdata.User;
```
테스트 코드에 production 코드와 동일하게 같은 클래스가 또 필요한 특별한 이유가 있었을까요?

eunsukko replied:
nextstep.mvc 를 프레임워크라고 생각해서 그렇게 사용했습니다..!
(그렇지않으면 slip에 있는 코드에 의존을 하게 됩니다. 프레임워크에서는 외부 코드를 몰라야한다고 생각했습니다)

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/handlermapping/AnnotationHandlerMapping.java:
```diff
@@ -43,48 +44,41 @@ public void initialize() {
+        try {
+            tryRegisterFromMethod(method, mappingBuilder);
+        } catch (RuntimeException e) {
+            log.info("error: {}", e);
```
error msg를 info 로그로 남겨도 괜찮을까요?

eunsukko replied:
앗 아니요...ㅠㅎ

프레임워크에서 로깅과 해당 프레임워크를 사용하는 곳에서의 로깅과 혼동이 있었습니다.

지금은 프레임워크에서는 레벨을 잘 나누어서 로깅을 해놓으면
사용하는 곳에서 적절히 정책을 적용(해당 모듈은 어떤 레벨로 어느 곳에 적을지)하는 책임을 가진다
이렇게 이해하려고 합니다.
(현재는 프레임워크와 slipp 코드 모두 slf4j 를 사용하기에, slipp 코드에서 로깅에 대한 처리를 재조정(어떤 레벨까지, 어느 곳에다가 써줄지)할 수 있는거라고 판단했습니다.)

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/handlermapping/HandlerExecutionFactory.java:
```diff
@@ -0,0 +1,66 @@
+                        return (ModelAndView) method.invoke(instance, request, response);
+                    } catch (IllegalAccessException | InvocationTargetException e) {
+                        log.error("should not be called: ", e);
+                        return null;
```
null을 return하는 것이 좋은 구조일까요?
또한 catch 하는 대신 throw를 하면 어떨까요?

eunsukko replied:
말씀해주신 부분을 고민해보았습니다.

null 이 발생하는 경우는 mvc 프레임워크 사용자가 프레임워크를 잘못사용할 경우입니다.
이 경우 프레임워크에서 할 수 있는 일은 사용자에게 최대한 빨리 알려주는 것이라고 생각이 들었습니다.
(콜백 내부에서 발생하는 예외는 사용자가 해당 `HandlerExecution`을 호출할 때 발생하는게 조금 아쉽긴 합니다 ㅠ!)

그래서 전체적으로 예외를 throw 하는 구조로 변경했습니다! (NextstepMvcException  이란 exception 으로 감싸서 throw 해줍니다.)

`HandlerExecutionFactory` 내부에서 checked Exception 을 처리하는 역할을 MethodWrapper 라는 내부 클래스를 둬서 위임해보았습니다. (현재는 이런 방식으로 Method 를 다룰 곳이 여기밖에 없다고 판단해서 내부 클래스로 두었습니다. 만들어보고 나니.. 다른 사람이 보기엔 더 불편할 수도 있지 않을까란 생각도 드네요..;)

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/handlermapping/HandlerKeyFactory.java:
```diff
@@ -0,0 +1,49 @@
+
+    public List<HandlerKey> fromMethod(Method method) {
+        RequestMapping requestMapping = method.getAnnotation(RequestMapping.class);
+        if (requestMapping == null) {
```
조건들을 더 명확하고 깔끔하게 정리할 수 있을 것 같아요!

코드 전반적으로 구조를 개선할 수 있는 부분이 있다고 생각하면, 고민해보고 깔끔하게 리팩토링도 같이 해보면 좋을 것 같아요 :)

eunsukko replied:
되도록 public 메소드에서는 해당 클래스에서 하는 역할이 보이도록 구성했습니다.
(로직 관련 부분을 숨겨서..  되도록 읽을 때 무엇을 하는지가 드러나도록 하였습니다.)

코드전반적으로도 살펴보았습니다..!
AnnotationHandlerMapping 부분이 확실히 간결해진 것 같습니다! 
(역할을 나눠보려고 시도했더니 지금의 형태가 되었네요..!)

jihan805 commented on slipp/src/main/java/slipp/controller/UserApiController.java:
```diff
@@ -0,0 +1,99 @@
+
+        ModelAndView modelAndView = new ModelAndView(JsonView.getInstance());
+        modelAndView.addObject("message", "created");
+        return new ModelAndView(JsonView.getInstance());
```
new ModelAndView(JsonView.getInstance()) 두번할 필요가 있을까요?

eunsukko replied:
잘못 구현되어있었네요.. 감사합니다!

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/handlermapping/AnnotationHandlerMapping.java:
```diff
@@ -43,48 +44,41 @@ public void initialize() {
-                return null;
-            }
-        };
+    private void tryRegisterFromMethod(Method method, Builder mappingBuilder) {
```
tryRegisterFromMethod 로 분리하지 않아도 괜찮을 것 같아요 :)

eunsukko replied:
요즘 try / catch 가 신경이 쓰여서... 기존 로직과 분리하려고 시도하다가 이렇게 되었습니다..

분리하는게 과하게 느껴질까요??

eunsukko replied:
리팩토링을 하다보니..
각 부분에서는 RuntimeException 을 발생시켜주고
DispatcherServlet 에서 한 번 감싸서 사용자에게 전달하는 식으로 변경되었습니다!
(CheckedException 을 되도록 빨리 RuntimeException 으로 변경하고
전체적으로 발생했던 RuntimeException 을 한 번 감싸서(NextstepMvcException으로 ) mvc 프레임워크 사용자에게 전달합니다)


이렇게 하면 사용자는 mvc 프레임워크에서 발생한 애러는 NextstepMvcException 으로 처리할 수 있고
상세 내용을 알고 싶으면 감싸진 예외를 확인하면 됩니다.
그리고 mvc 프레임워크 내부 코드에서는 try/catch 를 할 필요가 없어집니다. (NextstepMvcException 으로 감싸줄 한 군데가 필요한데.. 이는 현재 DispatcherServlet 에서 처리중입니다!)

jihan805 commented on nextstep-mvc/src/test/java/nextstep/mvc/handlermapping/HandlerExecutionFactoryTest.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    private Method findMethod(String methodName) {
+        return Arrays.asList(Clazz.class.getDeclaredMethods()).stream()
```
테스트 작성 👍 
테스트를 위한 메서드/클래스보다 실제 구현된 코드를 테스트할 수 있도록 해볼까요?

eunsukko replied:
실제 코드(slipp 에 존재하는) 클래스에 대해서 테스트해보라는 의미로 이해했습니다
(이 부분도 조금 고민인 부분은.. 프레임워크에서 slipp 의 코드를 알아야하는건가 입니다 ㅠ!)
그래서 간단하게 slipp 에 테스트 하나를 추가했는데.. 말씀하신 의도를 잘못 이해한 것 같기도 하네요ㅠㅎ

관련 테스트를 구현할 때.. `Clazz.class` 처럼 각 테스트에서  필요한 클래스를 추가해준 이유는.. `@Controller`나 `@RequestMapping` 등과 독립적이였으면 해서 였습니다.
(그렇다고 각 테스트마다 이렇게 데이터들을 만들면.. 읽기가 어려울까요?)


한편으로는
HandlerExecutionFactory 의 역할이 HandlerExecution 을 생성해주는 건데
,
테스트가 부실한 것도 같네요;; (실제 구현된 코드를 테스트하라고 하신게.. 생성된 HandlerExecution 을 실행해서 올바른 HandlerExecution 이 실행되는 것 까지 보라는 말씀이였을까요?)

jihan805 reviewed (changes requested):
구현 잘하셨네요 👍 
리뷰가 늦어져서 죄송합니다!
몇가지 피드백 남겼으니 확인해주세요 :)

eunsukko committed:
refactor: HandlerExecutionFactory Optional 반환 -> 예외

- 왜 변경(혹은 추가)이 있었나요?
	- Optional 을 반환해도 mvc에서는 처리해 줄 방법이 없음
	- 사용자에게 예외를 전달하는 것이 낫겠다는 판단
- 무엇이 변경되었나요?
	- mvc 전용 예외를 생성, 던지기

eunsukko committed:
refactor: JsonUtils 내부 메서드구조 리팩토링

- 왜 변경(혹은 추가)이 있었나요?
	- 기존 코드가 너무 string 에만 최적화되어 있음
- 무엇이 변경되었나요?
	- 각각의 OutputStream 에 대해서도 비슷한 방식으로 적용할 수 있도록 수정

eunsukko committed:
refactor: AnnotationHandlerMapping 리팩토링(내부 역할 분리)

- 왜 변경(혹은 추가)이 있었나요?
	- 내부 로직이 복잡했음
	- 여러가지 역할을 담당
- 무엇이 변경되었나요?
	- 독립된 역할을 기준으로 분리
		- 클래스를 얻는 역할
		- 메소드를 얻는 역할
		- 메소드를 등록하는 역할
	- 이전 단계에서 다음 단계에 필요한 정보를 모아서 제공

eunsukko committed:
refactor: HandlerExecutionFactory 테스트 추가

eunsukko committed:
feat: ReflectionTest 테스트 추가

- getField
	- 부모에 존재하는 접근가능한 필드도 읽어어는지 검증
- getMethod
	- 메소드 값 올바른지 검증

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/handler/handlermapping/HandlerExecutionFactory.java:
```diff
@@ -0,0 +1,96 @@
+        @Override
+        public Object newInstanceOfDeclaringClass() {
+            try {
+                return method.getDeclaringClass().getDeclaredConstructor().newInstance();
```
명확한 의도를 가지고 wrapper 클래스를 이용하여 처리👍 

다만, 프레임워크내 예외처리가 조금 더 일관성 있으면 좋을 것 같아요. 프레임워크 처럼 더 복잡한 구현의 경우 가독성도 중요하다고 생각해요.
추가로 
1. 에러 문구를 남길때 통일하여 남긴다 (깔끔하게 에러문구를 확인할 수 있겠죠?)
2. 위 return문의 chaining이 너무길 길다고 생각해요

jihan805 commented on nextstep-mvc/src/test/java/nextstep/mvc/handler/handlermapping/HandlerExecutionFactoryTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private Method findMethod(String methodName) {
+        return Arrays.asList(Clazz.class.getDeclaredMethods()).stream()
```
독립적으로 작성하는 방법은 좋은 것 같아요~!
다만 작성자가 보았을때도 복잡하거나 읽기 어려울 것 같다는 생각이 들면, 조금 개선이 필요하지 않을까요?
테스트 코드가 읽기 쉬워야 하는 이유에 대해서도 고민해보면 어떨까요?

jihan805 reviewed (approved):
안녕하세요~! 피드백 반영 잘해주셨네요 💯 
미션 마무리 하느라 고생 많았어요 :)
